### PR TITLE
backup: Add BACKUP INTO varient of BACKUP TO

### PR DESCRIPTION
**Pull request by dt**
The 'BACKUP ... TO x' syntax is used to write a backup _to_ x.

This change adds a new variation on that syntax with INTO instead of
TO, which indicates that the produced backup will be written to a path
chosen under x rather than directly at x. Effectively, this treats the
destination path a *collection* of BACKUPs to which we can add.

When BACKUP INTO x is run, it will pick a sub-path within x based
on the current time, similar to how automatically incremental backups already
do so within full backups when their path is passed to TO.

Additionally when a BACKUP started using INTO finished, the file x/LATEST_BACKUP
will be updated to contain the sub-path within x to which the backup was
written. This in turn can be used by the new shorthand 'BACKUP INTO LATEST IN x'
to append incremental changes to the most recent backup created in x.

This simplifies the task of taking periodic backups to just running 'BACKUP INTO x'
at whatever frequency one wants to backup. Additionally, if one wants to capture
changes incrementally, instead of needing to manually track where to append them,
one can just run 'BACKUP INTO LATEST IN x' at the desired change-capture frequency.
Doing so in addition to running the aforementioned 'BACKUP INTO x' at the desired
full-backup cadence also has the advantage that it will continue to append to the most
recent _completed_ backup until the next one finishes and updates the LATEST_BACKUP
file, avoiding the "RPO blip" that can sometimes arise in other schedulings of full
vs incremental backups.

Release note (enterprise change): add 'BACKUP ... INTO' alternative to 'BACKUP TO' which automatically picks a new path within the destination.

### CODE REVIEW

**Review by miretskiy — CHANGES_REQUESTED** [review]
> Reviewed 16 of 16 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52329)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @pbardea)

___
*[pkg/ccl/backupccl/backup_job.go, line 531 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDtrjtt4WMPlFop-a5i:-MDtrjtt4WMPlFop-a5j:b-145bu7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2117e1825c11ffd233b72d8a04cb48a9c720e7/pkg/ccl/backupccl/backup_job.go#L531)):*
> ```Go
> 		defer exportStore.Close()
> 		if err := exportStore.WriteFile(ctx, latestFileName, strings.NewReader(suffix)); err != nil {
> 			return err
> ```

I wonder what we can/should do in this case.
There are few issues I can see:
  1. This error return, of course, fails the entire backup, even though we just finished it successfully.
  2. It is rather inconsistent with how we handle other actions we do after backup() -- such failure to release TS just logs an error, failure to delete the checkpoint just logs a warning. 

This `return error` is not a "new issue" per say -- just magnifies the fact that we don't really have a way of handling "backup finished fine, but we couldn't do some extra cleanup steps correctly".
___
*[pkg/ccl/backupccl/backup_planning.go, line 521 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDtiPEb6ixs9AATn4AD:-MDtiPEc-9p_bU1TXW64:b-3i4417) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2117e1825c11ffd233b72d8a04cb48a9c720e7/pkg/ccl/backupccl/backup_planning.go#L521)):*
> ```Go
> 		// the latest file.
> 		var collectionPath string
> 		if backupStmt.Nested {
> ```

nit: Can we move this entire block to a helper?
I just think

```
collectionPath := getCollectionPath(....) 
```

is a lot easier to read (and test).
___
*[pkg/ccl/backupccl/backup_planning.go, line 556 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDtj-X_9BSARnm_0jUf:-MDtj-X_9BSARnm_0jUg:b-wdjpbb) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2117e1825c11ffd233b72d8a04cb48a9c720e7/pkg/ccl/backupccl/backup_planning.go#L556)):*
> ```Go
> 					collectionPath = u.Path
> 				}
> 				u.Path = path.Join(u.Path, chosenSuffix)
> ```

do we want to verify that chosenSuffix doesn't have any funny characters? (../../)?

Do you think it might make sense to just store the "end time" of the latest backup, and just try to parse here?

I see that u.Path existence is checked below... Perhaps a quick comment to that effect?
___
*[pkg/jobs/jobspb/jobs.proto, line 90 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDtfgy9-QkxMrbnrYxV:-MDtfgy9-QkxMrbnrYxW:b8a9m4q) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2117e1825c11ffd233b72d8a04cb48a9c720e7/pkg/jobs/jobspb/jobs.proto#L90)):*
> ```
>   // for nesting. Thus replacing the path in the above URI with this path would
>   // yield the original user supplied collection URI.
> ```
I think i understand what you mean; it's just this proto has few URI fields, so this comment is a bit hard to parse.  If I understand correctly, if this is set, then this field is a "basepath" and all other URIs are "basenames"?
___
*[pkg/sql/parser/parse_test.go, line 1481 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDtfB1T33MbuvdY1UM6:-MDtfB1T33MbuvdY1UM7:b-vmwll2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2117e1825c11ffd233b72d8a04cb48a9c720e7/pkg/sql/parser/parse_test.go#L1481)):*
> ```Go
> 
> 		{`BACKUP TABLE foo TO 'bar'`},
> 		{`BACKUP TABLE foo INTO 'bar'`},
> ```

I think you need to add few more tests, including the latest keyword, with/without other options;
plus the restore test as well.
___
*[pkg/sql/parser/sql.y, line 2064 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDteF-R3If4-wMhju-i:-MDteF-S5DiK7IYXryRZ:b-ew23r2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2117e1825c11ffd233b72d8a04cb48a9c720e7/pkg/sql/parser/sql.y#L2064)):*
> ```Yacc
>     $$.val = &tree.Backup{
>       Targets: $2.targetListPtr(), To: $4.stringOrPlaceholderOptList(),
>       AsOf: $5.asOfClause(), Options: *$6.backupOptions(),
> ```

nit: personally, i prefer each option on separate line (when we have many options)

**Review by pbardea — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52329)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @miretskiy)

___
*[pkg/ccl/backupccl/backup_job.go, line 531 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDtrjtt4WMPlFop-a5i:-MDu0nNK8j2fV19Qwvoj:b-p3rad4) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2117e1825c11ffd233b72d8a04cb48a9c720e7/pkg/ccl/backupccl/backup_job.go#L531)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I wonder what we can/should do in this case.
There are few issues I can see:
  1. This error return, of course, fails the entire backup, even though we just finished it successfully.
  2. It is rather inconsistent with how we handle other actions we do after backup() -- such failure to release TS just logs an error, failure to delete the checkpoint just logs a warning. 

This `return error` is not a "new issue" per say -- just magnifies the fact that we don't really have a way of handling "backup finished fine, but we couldn't do some extra cleanup steps correctly".

</blockquote></details>

My view is that we have not backed up successfully until we have written all metadata as well as all the actual data. I would prefer to alert the user that we couldn't write everything we wanted, since future BACKUPs may otherwise behave in an unexpected way if users aren't looking for the appropriate logs.

One feature that I think would help alleviate this issue would be to add the ability to retry some failed BACKUPS. Restarting the entire backup because we couldn't update the LATEST_BACKUP file is a pretty big ask.
___
*[pkg/ccl/backupccl/backup_planning.go, line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDtvZ9I0p2kJ1Pv2LNk:-MDtvZ9K03E0czZgDyse:b-ks81qr) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2117e1825c11ffd233b72d8a04cb48a9c720e7/pkg/ccl/backupccl/backup_planning.go#L62)):*
> ```Go
> 	defaultLocalityValue     = "default"
> 	dateBasedFolderName      = "/20060102/150405.00"
> 	latestFileName           = "LATEST"
> ```

I think it might make sense to move these to `manifest_handling.go` with the other file consts.

Also, the comments and commit call this file `LATEST_BACKUP`, but here it's just `LATEST`.

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52329)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @miretskiy)

___
*[pkg/ccl/backupccl/backup_job.go, line 531 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDtrjtt4WMPlFop-a5i:-MDttZeS8olND1FNuT8p:bc89wki) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2117e1825c11ffd233b72d8a04cb48a9c720e7/pkg/ccl/backupccl/backup_job.go#L531)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

My view is that we have not backed up successfully until we have written all metadata as well as all the actual data. I would prefer to alert the user that we couldn't write everything we wanted, since future BACKUPs may otherwise behave in an unexpected way if users aren't looking for the appropriate logs.

One feature that I think would help alleviate this issue would be to add the ability to retry some failed BACKUPS. Restarting the entire backup because we couldn't update the LATEST_BACKUP file is a pretty big ask.
</blockquote></details>

Ditto --  this isn't an extra cleanup or otherwise something optional -- updating latest is *required* for this backup to "finish successfully" as we define it since that includes becoming the new "latest" backup. If we failed to do that, then we can't report "success".

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52329)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, and @pbardea)

___
*[pkg/ccl/backupccl/backup_planning.go, line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDtvZ9I0p2kJ1Pv2LNk:-MDuilLv6PO1hgJM82k1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2117e1825c11ffd233b72d8a04cb48a9c720e7/pkg/ccl/backupccl/backup_planning.go#L62)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

I think it might make sense to move these to `manifest_handling.go` with the other file consts.

Also, the comments and commit call this file `LATEST_BACKUP`, but here it's just `LATEST`.
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/backup_planning.go, line 521 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDtiPEb6ixs9AATn4AD:-MDuu7UrAYBvntpCNutO:b400w10) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2117e1825c11ffd233b72d8a04cb48a9c720e7/pkg/ccl/backupccl/backup_planning.go#L521)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

nit: Can we move this entire block to a helper?
I just think

```
collectionPath := getCollectionPath(....) 
```

is a lot easier to read (and test).

</blockquote></details>

hmm, tried this but i'm not sure i like it. partly because half the lines in this block are referencing vars in the local scope, so it needs a ton of args, but more importantly, it because it has side-effects: it rewrites `to` in a conditional branch. 

I think I see a potential refactor that pulls this and the big block below were we also re-write `to` or `defaultURIs` for appending into a "resolveBackupDestination` helper, but I'd rather do that separately, rather than just pulling out these lines and making it harder, IMO, to follow how we arrive at `to` and `defaultURI`.

I did do a tiny bit of tidying though so there's less in this block.
___
*[pkg/ccl/backupccl/backup_planning.go, line 556 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDtj-X_9BSARnm_0jUf:-MDuuifTDBhPWkMqBgqw:b326v9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2117e1825c11ffd233b72d8a04cb48a9c720e7/pkg/ccl/backupccl/backup_planning.go#L556)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

do we want to verify that chosenSuffix doesn't have any funny characters? (../../)?

Do you think it might make sense to just store the "end time" of the latest backup, and just try to parse here?

I see that u.Path existence is checked below... Perhaps a quick comment to that effect?
</blockquote></details>

hmm, I'm not sure. if it has `../` (it won't, but just if it did e.g. because you edited LATEST by hand) then i'd say fine -- `path.Join` will flatten it, and we'll just go look in the resulting path. If the resulting path doesn't point at valid backup,  well you'll get an error, but that's no different than if you edited it to point to some random non-backup path without using ../, so I don't think there's anything special to consider here. 

(Note: if you were worried about someone "escaping" or something, _that_ is up to the storage impl to decide where we can or cannot go, so that isn't something we should check here or other rando places. )
___
*[pkg/sql/parser/parse_test.go, line 1481 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDtfB1T33MbuvdY1UM6:-MDuuvvm8C51v6UBzA1B:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2117e1825c11ffd233b72d8a04cb48a9c720e7/pkg/sql/parser/parse_test.go#L1481)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think you need to add few more tests, including the latest keyword, with/without other options;
plus the restore test as well.
</blockquote></details>

Done.

**Review by pbardea — COMMENTED** [review]
Generally LGTM - mostly minor comments. I think this will make managing backups a lot easier!

Is there a docs ticket filed for this already? I'm not 100% clear on when we create that ticket for tracking the relevant docs updates.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-:-MDz2GHYCPz1gfRUJTQY:btr861w)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @miretskiy)

___
*[pkg/ccl/backupccl/backup_planning.go, line 521 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDtiPEb6ixs9AATn4AD:-MDyzIaD07XBImqR1hng:b-1e8he8) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2117e1825c11ffd233b72d8a04cb48a9c720e7/pkg/ccl/backupccl/backup_planning.go#L521)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

hmm, tried this but i'm not sure i like it. partly because half the lines in this block are referencing vars in the local scope, so it needs a ton of args, but more importantly, it because it has side-effects: it rewrites `to` in a conditional branch. 

I think I see a potential refactor that pulls this and the big block below were we also re-write `to` or `defaultURIs` for appending into a "resolveBackupDestination` helper, but I'd rather do that separately, rather than just pulling out these lines and making it harder, IMO, to follow how we arrive at `to` and `defaultURI`.

I did do a tiny bit of tidying though so there's less in this block.
</blockquote></details>

Just as a side-note, I think such a refactor would be beneficial. I've been trying to split up this code into more helpers, but frequently run into the issues you've mentioned.
___
*[pkg/ccl/backupccl/backup_planning.go, line 580 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDz2Z_9-VsTSvj1sECR:-MDz2Z_9-VsTSvj1sECS:b-iaiubc) ([raw file](https://github.com/cockroachdb/cockroach/blob/df2e326f896bf8950539a96119b867874ce36c91/pkg/ccl/backupccl/backup_planning.go#L580)):*
> ```Go
> 		}
> 
> 		// chosenSuffix is automaically chosen sffix with a collection path if we're
> ```

nit: sffix
___
*[pkg/ccl/backupccl/backup_planning.go, line 677 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDz-VLD-y0jDdLzTR19:-MDz-VLE2hbKba1Qymws:b-gw5kob) ([raw file](https://github.com/cockroachdb/cockroach/blob/df2e326f896bf8950539a96119b867874ce36c91/pkg/ccl/backupccl/backup_planning.go#L677)):*
> ```Go
> 				return err
> 			}
> 			defer defaultStore.Close()
> ```

Don't we need to wrap this Close() call in a closure since we'll overwrite `defaultStore` in the same func?
___
*[pkg/ccl/backupccl/backup_test.go, line 290 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDz04022fKt1z7VW5BU:-MDz04022fKt1z7VW5BV:b-qdqdq3) ([raw file](https://github.com/cockroachdb/cockroach/blob/df2e326f896bf8950539a96119b867874ce36c91/pkg/ccl/backupccl/backup_test.go#L290)):*
> ```Go
> 	sqlDB.Exec(t, "ALTER TABLE data.bank RENAME TO data.renamed")
> 	sqlDB.Exec(t, "BACKUP TO ($1, $2, $3)", backups...)
> 	sqlDB.Exec(t, "BACKUP INTO LATEST IN ($1, $2, $3)", collections...)
> ```

Should we add a test case where we build on top of an incremental (2 `INTO LATEST` in a row) to ensure that an incremental doesn't muck up the LATEST_BACKUP file?
___
*[pkg/ccl/backupccl/manifest_handling.go, line 60 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDz0hMP2yK5EgMpi78T:-MDz0hMP2yK5EgMpi78U:br35c2u) ([raw file](https://github.com/cockroachdb/cockroach/blob/df2e326f896bf8950539a96119b867874ce36c91/pkg/ccl/backupccl/manifest_handling.go#L60)):*
> ```Go
> 
> 	dateBasedFolderName = "/20060102/150405.00"
> 	latestFileName      = "LATEST"
> ```

Minor nit: the commit message could also use a s/LATEST_BACKUP/LATEST

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52329)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, and @pbardea)

___
*[pkg/ccl/backupccl/backup_planning.go, line 521 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDtiPEb6ixs9AATn4AD:-MDz6OhQ3oNNKGTaaEGa:bgx13me) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2117e1825c11ffd233b72d8a04cb48a9c720e7/pkg/ccl/backupccl/backup_planning.go#L521)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Just as a side-note, I think such a refactor would be beneficial. I've been trying to split up this code into more helpers, but frequently run into the issues you've mentioned.
</blockquote></details>

i played with this a little again last night and i think i see a way forward, but i don't think i want to pull it into this change.

I think we re-work the existing block below that re-writes `to` / `defaultURI` to _not_ also load `priorBackups` at the same time. Then we also do not need to load encryption options in the middle of all this, in duplicated blocks.

We put all that in a "resolveDestination" helper that returns default/localityURIs and optional inc-base URI. 
Then we  crypto config, once, using that inc-base if set, instead of in each branch. 
_Then_ we load prior backups from inc-base.

I think that'll clean this up, but it a lot more churn than i want in a new-feature PR if i can avoid it.
___
*[pkg/ccl/backupccl/backup_planning.go, line 677 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDz-VLD-y0jDdLzTR19:-MDz4YRRAO3qVygqleMB:bqhlmor) ([raw file](https://github.com/cockroachdb/cockroach/blob/df2e326f896bf8950539a96119b867874ce36c91/pkg/ccl/backupccl/backup_planning.go#L677)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Don't we need to wrap this Close() call in a closure since we'll overwrite `defaultStore` in the same func?
</blockquote></details>

I thought I did away with over-writing it (specifically to avoid this)?
___
*[pkg/ccl/backupccl/backup_planning.go, line 969 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDz55So8_Je2Xa61zUX:-MDz55So8_Je2Xa61zUY:b-o1yo26) ([raw file](https://github.com/cockroachdb/cockroach/blob/df2e326f896bf8950539a96119b867874ce36c91/pkg/ccl/backupccl/backup_planning.go#L969)):*
> ```Go
> 		}
> 
> 		defaultStore, err := makeCloudStorage(ctx, defaultURI, p.User())
> ```

nb: this is `:=`, not `=`

**Review by pbardea — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52329)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, and @pbardea)

___
*[pkg/ccl/backupccl/backup_planning.go, line 677 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDz-VLD-y0jDdLzTR19:-MDzD1e47vLtWq-I19OJ:b-x30d0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/df2e326f896bf8950539a96119b867874ce36c91/pkg/ccl/backupccl/backup_planning.go#L677)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I thought I did away with over-writing it (specifically to avoid this)?
</blockquote></details>

Ah, indeed. Missed the nesting.

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52329)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @pbardea)

___
*[pkg/ccl/backupccl/backup_test.go, line 290 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDz04022fKt1z7VW5BU:-MDzJ5Jd2R7s4KQ38F26:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df2e326f896bf8950539a96119b867874ce36c91/pkg/ccl/backupccl/backup_test.go#L290)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Should we add a test case where we build on top of an incremental (2 `INTO LATEST` in a row) to ensure that an incremental doesn't muck up the LATEST_BACKUP file?
</blockquote></details>

Done.
___
*[pkg/jobs/jobspb/jobs.proto, line 90 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDtfgy9-QkxMrbnrYxV:-MDzJAX2ErPwL14qWW5m:b-sg0ohj) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2117e1825c11ffd233b72d8a04cb48a9c720e7/pkg/jobs/jobspb/jobs.proto#L90)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> ```
>   // for nesting. Thus replacing the path in the above URI with this path would
>   // yield the original user supplied collection URI.
> ```
I think i understand what you mean; it's just this proto has few URI fields, so this comment is a bit hard to parse.  If I understand correctly, if this is set, then this field is a "basepath" and all other URIs are "basenames"?
</blockquote></details>

I changed this field to be the whole collection URI since I think that makes it conceptually a bit simpler.

**Review by pbardea — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52329)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @miretskiy)

___
*[pkg/ccl/backupccl/backup_planning.go, line 521 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52329#-MDtiPEb6ixs9AATn4AD:-ME-A8SC1e6OVVlZQ92K:b3uh2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c2117e1825c11ffd233b72d8a04cb48a9c720e7/pkg/ccl/backupccl/backup_planning.go#L521)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

i played with this a little again last night and i think i see a way forward, but i don't think i want to pull it into this change.

I think we re-work the existing block below that re-writes `to` / `defaultURI` to _not_ also load `priorBackups` at the same time. Then we also do not need to load encryption options in the middle of all this, in duplicated blocks.

We put all that in a "resolveDestination" helper that returns default/localityURIs and optional inc-base URI. 
Then we  crypto config, once, using that inc-base if set, instead of in each branch. 
_Then_ we load prior backups from inc-base.

I think that'll clean this up, but it a lot more churn than i want in a new-feature PR if i can avoid it.
</blockquote></details>

Yep, agreed that it's a lot of code movement. Optionally, perhaps a TODO around the block below would be beneficial to help ensure that the refactor doesn't get lost?

**Comment by dt** [comment]
TFTR! 

bors r+
